### PR TITLE
Adds framework for status effects

### PR DESCRIPTION
**Pull request by ghost**
Added framework for status effects with lots of customization, including duration, effects, and pretty much everything else. Files are located in `__DEFINES/status_effects.dm`, and `datum/status_effects`. There are two procs for adding and removing them, both at the mob level - `apply_status_effect` and `remove_status_effect`, both of which take each effect's path or define.

This allows for making timed or constant effects without having random loops all over the place. For instance, if you want to have a creature mark its target to die after one minute, you can make a status effect that lasts one minute and calls `death()` on the owner once it expires.

Code suggestions and criticisms are appreciated and encouraged.

### CODE REVIEW

File: code/datums/status_effects/status_effect.dm [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+	var/cosmetic = FALSE //If the status effect only exists for flavor.
+
+/datum/status_effect/New()
+	..()
+	spawn(1) //Give us time to set any variables
+		start_ticking()
+
+/datum/status_effect/proc/start_ticking()
+	on_apply()
+	while(TRUE) //please don't kill me :c
```
Comment (Xhuis) [inline code/datums/status_effects/status_effect.dm]: I found that this increased the flexibility I have, and it shouldn't have any recursive effects due to it being an individual datum.

Comment (RemieRichards) [inline code/datums/status_effects/status_effect.dm]: it should just be something that processes, not custom snowflake.

Comment (Xhuis) [inline code/datums/status_effects/status_effect.dm]: I considered that. Would a new subsystem be worth making? I used SSobj but its delay for processing really hindered me.

Comment (RemieRichards) [inline code/datums/status_effects/status_effect.dm]: just put it on the fast processing SS?

Comment (Xhuis) [inline code/datums/status_effects/status_effect.dm]: Could be doable. Fastprocess ticks five times per second, right?

File: code/datums/status_effects/status_effect.dm [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+//////////////////
+// HELPER PROCS //
+//////////////////
+
+/mob/living/proc/apply_status_effect(effect)
+	var/datum/status_effect/S = new effect
+	S.owner = src
+
+/mob/living/proc/remove_status_effect(effect)
+	for(var/datum/status_effect/S in status_effects)
```
Comment (RemieRichards) [inline code/datums/status_effects/status_effect.dm]: so you can have multiples of the same buff? that's... weird.
Like sure, buff-stacking to get a better version of the buff, but just plain having two copies? weird.

Comment (Xhuis) [inline code/datums/status_effects/status_effect.dm]: That's an oversight! I'll make a variable for it.

**Comment by optimumtact** [comment]
Are you planning on moving the fire and atmos status effects to this system?

i.e being set on fire applies the burning effect (alert and damage) and losing air or breathing poison applies the atmos alerts and associated damage?

It's good to have a system as in the past people have often used the chem or disease systems to apply these effects instead.

If someone was enterprising they could also overhaul disease symptoms to be effects, so diseases would be a layer above this system dealing with the virology aspects.

**Comment by Xhuis** [comment]
Disease already has its own datum thingy, so I probably won't change that. On fire effects would be easy enough to do, but I think this will be a more niche system.

**Comment by optimumtact** [comment]
Why would it be niche? There are many items in the game that need to apply a consistent/timed status effect on the bearer or on some other condition.

**Comment by Xhuis** [comment]
I'm sure there's plenty of ways this can be applied. I just don't know if things like atmos and viruses would be suitable candidates as they already have lots of code written around them. I don't want _really_ complex stuff - only small stuff like "take this amount of damage over this amount of time" and stuff like that. If you can list off stuff that I can refactor to status effects, I'd definitely refactor them... tomorrow. Sleep now.

File: code/datums/status_effects/status_effect.dm (hunk 1 above)
Comment (optimumtact) [inline code/datums/status_effects/status_effect.dm]: We're very unlikely to merge this when it reimplements the process scheduling and subsystem ticker we already have for it's own benefit.

**Comment by WJohn** [comment]
Weaken and fire and weightlessness and sleep or whatever should totally be moved to status effect with alerts. More obvious feedback for what's going on.

**Comment by AnturK** [comment]
Good idea but needs to actually standarize current statuses. Also ideally link it with the popup alerts.

File: code/datums/status_effects/status_effect.dm (hunk 1 above)
Comment (AnturK) [inline code/datums/status_effects/status_effect.dm]: If this is meant to be standard effect system separate controller is reasonable.

**Comment by Xhuis** [comment]
* Moved Invoke Inath-neq to a status effect, Inath-neq's Endowment. More refactors to come.
* Made status effects throw alerts to the afflicted' screen, based on icons found in `status_effects.dmi`.
* Refactored fastprocess into a processing subsystem that fires once per second. Fastprocess is a child of this that fires at its previous rate.
* Added support for non-stacking status effects, which all status effects will utilize by default.
* Moved all processing subsystems into `/subsystems/processing/`.

**Comment by optimumtact** [comment]
Ready for review?

**Comment by Xhuis** [comment]
Yeah.

File: code/controllers/subsystem/processing/processing.dm [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+	//cache for sanic speed (lists are references anyways)
+	var/list/current_run = currentrun
+
+	while(current_run.len)
+		var/datum/thing = current_run[current_run.len]
+		current_run.len--
+		if(thing)
+			thing.process(wait)
+		else
+			SSfastprocess.processing -= thing
```
Comment (RemieRichards) [inline code/controllers/subsystem/processing/processing.dm]: error?

File: code/datums/status_effects/status_effect.dm [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+/datum/status_effect/proc/tick() //Called every tick.
+/datum/status_effect/proc/on_remove() //Called whenever the buff expires or is removed.
+
+//////////////////
+// HELPER PROCS //
+//////////////////
+
+/mob/living/proc/apply_status_effect(effect)
+	var/datum/status_effect/S = new effect
+	for(var/datum/status_effect/S2 in status_effects)
```
Comment (RemieRichards) [inline code/datums/status_effects/status_effect.dm]: you could just make ```status_effects``` an assoc list of ```type = status_effect instance```, that prevents dupes, is faster to access, and if you choose allows setting an effect again to essentially just "refresh" it.

File: code/controllers/subsystem/processing/processing.dm (hunk 3 above)
Comment (Xhuis) [inline code/controllers/subsystem/processing/processing.dm]: Yes, thank you

File: code/datums/status_effects/status_effect.dm (hunk 4 above)
Comment (Xhuis) [inline code/datums/status_effects/status_effect.dm]: So each status effect would exist at roundstart and be "assigned" to mobs who lose it after the duration, but the effect still exists? I suppose that could work.

Comment (RemieRichards) [inline code/datums/status_effects/status_effect.dm]: the list wouldn't have to start with anything. if something tried to remove a non-existent effect, there'd be no issues, no boom, no runtime explosion. the list would start empty, but would slowly be filled like an assoc list.

it's all just better code-bookkeeping.

Comment (Xhuis) [inline code/datums/status_effects/status_effect.dm]: Ah, I see what you mean! Good point.

**Comment by optimumtact** [comment]
Any chance of working on those suggestions this week @Xhuis ?

**Comment by Xhuis** [comment]
Sure. I'll do some work when I get the opportunity.

File: code/datums/status_effects/status_effect.dm (hunk 4 above)
Comment (Xhuis) [inline code/datums/status_effects/status_effect.dm]: I've been working on this but I'm pretty atrocious with associative list syntax and I wanted to know what the problem with this would be.

Here's what I have for the `New()` for status effects:
```
/datum/status_effect/New()
	..()
	spawn(1) //Give us time to set any variables
		start_ticking()
		status_effects += src["[owner.real_name][id]"]
```
But I can't help but feel that this is wrong. Am I missing something?

Comment (RemieRichards) [inline code/datums/status_effects/status_effect.dm]: yes that is indeed wrong.
```status_effects``` is the assoc list, so you do the list operation on IT.
```status_effects["[owner.real_name][id]"] = src```, and it's ```=``` because you're setting the value of ```status_effects["[owner.real_name][id]"]```, as opposed to adding to that value (if that value was itself a list, for example.

tl;dr
```
list[1] = "A" //First element is now "A"
assoc_list["some_key"] = "A" //the key "some_key" now points to "A"
```

Comment (Xhuis) [inline code/datums/status_effects/status_effect.dm]: Thanks. I'll post again if I need further help.

Comment (Xhuis) [inline code/datums/status_effects/status_effect.dm]: Hey, reporting back in. I've refactored it for the most part, but there's a small obstacle I can't surmount. It's with this code:
```
/mob/living/proc/apply_status_effect(datum/status_effect/effect)
	if(status_effects["[real_name][initial(effect.id)]"])
		return 0
	var/datum/status_effect/S = new effect
	S.owner = src
	return 1
```
It works fine, but removing won't. Is typecasting `effect` as an actual status effect correct, and will `initial(effect.id)` work in this context?

Comment (RemieRichards) [inline code/datums/status_effects/status_effect.dm]: yes, typecasting like that works fine.
it's a little dangerous because you're not checking if effect is a path or not, but it works.

Comment (Xhuis) [inline code/datums/status_effects/status_effect.dm]: All right. My issue here is that `apply_status_effect` always returns 1, meaning that it never picks up the status effect in question. This is included in the `New()` for status effects now.
`status_effects["[owner.real_name][id]"] = src`
So I'm not entirely sure what's going wrong. I think it's because `apply_status_effect` takes a path, not an actual object. Would it be better to just make a new status effect for reference and delete it once the code is done with it?

Comment (RemieRichards) [inline code/datums/status_effects/status_effect.dm]: @Xhuis you're accessing the id from the path using ```initial()``` so it doesn't matter that it's not an object.

does your ```apply_status_effect()``` look exactly as that comment? because you're setting S's owner AFTER ```New()```, when you just said you put the assoc list addition in ```New()```

Comment (Xhuis) [inline code/datums/status_effects/status_effect.dm]: Still having issues after some tweaks, so I'll post the code to hear your own thoughts. Here's everything relevant.

```
/datum/status_effect/New(loc, mob/living/initial_owner)
	..()
	if(initial_owner)
		owner = initial_owner
	status_effects["[owner.real_name][id]"] = src
	start_ticking()

/datum/status_effect/Destroy()
	status_effects["[owner.real_name][id]"] = null
	..()

/mob/living/proc/apply_status_effect(datum/status_effect/effect)
    if(status_effects["[real_name][initial(effect.id)]"])
        return 0
    var/datum/status_effect/S = new effect
    S.owner = src
    return 1

/mob/living/proc/remove_status_effect(datum/status_effect/effect)
	var/datum/status_effect/S = status_effects["[real_name][initial(effect.id)]"]
	if(!S)
		return 0
	world << S.name
	qdel(S)
	return 1
```

Comment (RemieRichards) [inline code/datums/status_effects/status_effect.dm]: Here's a better ```New()```:
```
/datum/status_effect/New(new_owner)
    ..()
    if(new_owner)
        owner = new_owner
        status_effects["[owner.real_name][id]"] = src
        start_ticking()
    else
        qdel(src)
```
Better ```Destroy()```:
```
/datum/status_effect/Destroy()
    status_effects -= "[owner.real_name][id]" 
    return ..() //Destroy()s MUST return their hint.
```

Comment (Xhuis) [inline code/datums/status_effects/status_effect.dm]: Thanks again for the help. I'll stay in touch.

Comment (Xhuis) [inline code/datums/status_effects/status_effect.dm]: Afraid there's no change. It still adds duplicate effects and fails to remove existing ones.

Comment (RemieRichards) [inline code/datums/status_effects/status_effect.dm]: use the person's ```\ref[]``` (```\ref[src]```) instead of their ```real_name```

Comment (Xhuis) [inline code/datums/status_effects/status_effect.dm]: That just throws a bunch of undefined variable errors.

Comment (Xhuis) [inline code/datums/status_effects/status_effect.dm]: Losing motivation because this _still_ won't work. Reverting back to the old system.

Comment (RemieRichards) [inline code/datums/status_effects/status_effect.dm]: are you sure you used ```\ref[src]``` right? it's ```"some text blah blah \ref[a_var_such_as_src]"```, where as I'm assuming you thought to add ```[]``` around the whole statement, which is wrong.

Comment (Xhuis) [inline code/datums/status_effects/status_effect.dm]: So the syntax would be:
`status_effects[\ref[owner][id]]`?

Comment (RemieRichards) [inline code/datums/status_effects/status_effect.dm]: ```status_effects["\ref[owner][id]"]```

Comment (Xhuis) [inline code/datums/status_effects/status_effect.dm]: Through debugging, I discovered that `initial(effect.id)` when referring to the path comes up null. I think it _does_ have to be called on an object.

Comment (RemieRichards) [inline code/datums/status_effects/status_effect.dm]: that makes no sense, as it's never null, at worst it's "effect"

Comment (Xhuis) [inline code/datums/status_effects/status_effect.dm]: Yeah, I don't know either. I think the answer boils down to "Byond".

Comment (optimumtact) [inline code/datums/status_effects/status_effect.dm]: why are you doing intial(id)?

**Comment by optimumtact** [comment]
Is this ready to go?

**Comment by Xhuis** [comment]
No. It still doesn't work and I'm trying to figure out the best solution for the problem it's having.

**Comment by Razharas** [comment]
Make crit into a status effect

**Comment by Xhuis** [comment]
@Razharas That'd be for another PR since it would involve changing the fundamentals of human `Life` code.

Anyway, we have two options with this PR;
1. Use the inferior, slower, and clunkier method and merge it since the other method doesn't work
2. Close it or put it on hiatus until I can figure out _why_ it doesn't work

I've tried getting the ideal method to work for almost a week now, to no avail. What do you think, maintainers?

**Comment by optimumtact** [comment]
I wouldn't want to hold up what is a good overhaul for just that, so go back to the slower method for now and we can look at fixing it later.

**Comment by Xhuis** [comment]
All right, once I finish that highlander PR I'll fix up this one.

**Comment by Xhuis** [comment]
I reverted to the old system. It's wonky and _barely_ functional, but it works for the time being and doesn't cause any gameplay issues. I'll post back once I've done some more testing.

**Comment by Xhuis** [comment]
* Inath-neq's Endowment has an icon now and properly renders its text.

This is merge-ready!

**Comment by Iamgoofball** [comment]
@Xhuis you didnt fucking add a status effect removal proc

**Comment by Iamgoofball** [comment]
or a "has status effect" proc

**Comment by Xhuis** [comment]
Actually, I did, but they refused to work so I implemented the barebones version to work on later.
